File: java/triangle/src/main/java/Triangle.java
public class Triangle {

  private final double sideA;
  private final double sideB;
  private final double sideC;

  public Triangle(double sideA, double sideB, double sideC) throws TriangleException {
    if (triangleInvalid(sideA, sideB, sideC)) { throw new TriangleException(); }

    this.sideA = sideA;
    this.sideB = sideB;
    this.sideC = sideC;
  }

  public TriangleKind getKind() {
    if (sideA == sideB && sideB == sideC && sideA == sideC) {
      return TriangleKind.EQUILATERAL;
    } else if (sideA == sideB || sideA == sideC || sideB == sideC) {
      return TriangleKind.ISOSCELES;
    } else {
      return TriangleKind.SCALENE;
    }
  }

  private boolean triangleInvalid(double sideA, double sideB, double sideC) {
    return triangleInequality(sideA, sideB, sideC) ||
           triangleInequality(sideC, sideB, sideA) ||
           triangleInequality(sideA, sideC, sideB);
  }

  private boolean triangleInequality(double side1, double side2, double side3) {
    return (side1 + side2 <= side3);
  }

}

File: ruby/exercises/triangle/triangle.rb
# Triangle
class Triangle
  attr_accessor :side_a, :side_b, :side_c
  def initialize(side_a, side_b, side_c)
    @side_a = side_a
    @side_b = side_b
    @side_c = side_c
  end

  def kind
    raise TriangleError if error?
    return :equilateral if equilateral?
    return :isosceles   if isosceles?
    :scalene
  end

  private

  def error?
    triangle_inequality?(side_a, side_b, side_c) ||
      triangle_inequality?(side_b, side_c, side_a) ||
      triangle_inequality?(side_a, side_c, side_b)
  end

  def triangle_inequality?(side_x, side_y, side_z)
    side_x + side_y <= side_z
  end

  def equilateral?
    side_a == side_b && side_b == side_c
  end

  def isosceles?
    side_a == side_b || side_b == side_c || side_a == side_c
  end
end

TriangleError = Class.new(StandardError)
